CODEBASE: Expand lint rules, and Aliases are stored as maps

Snarling opened the pull request:
As mentioned in #474, the expected lint rules were not being applied. This PR is to enable the desired lint rules.

So far ada0e13 restores the `eslint:recommended` rules and makes the appropriate codebase changes to comply with the additional rules. Most of the violations were direct usage of hasOwnProperty from an object, so I created a helper function that instead calls hasOwnProperty from the Object prototype and replaced all usage. (also see 94fc8d6)

151daf0 contains mostly fixes the linter is able to automatically apply with `@typescript-eslint/recommended-requiring-type-checking` enabled. Most of these changes are removed type assertions that were unnecessary due to that type already being guaranteed.

94fc8d6 Switches out the custom helper hasOwnProps with Object.hasOwn

7dacaaa Changes aliases to use a map rather than a plain object. This streamlines usage and is just a better data structure for what it is. Worth noting that prior to this PR, there was a bug where setting an alias with the name hasOwnProperty would cause a crash, though that was already addressed by previous fixes.

Snarling committed:
Add eslint:recommended lint rules

Snarling committed:
Auto-lint changes

From temporarily enabling `@typescript-eslint/recommended-requiring-type-checking"` rules.

There are 936 other violations of these rules that could not be auto-fixed. Will start working through those on next commit.

Also the auto-lint tried to remove type assertions for IPAddress and ScriptKey since they were just string, so I enabled the previously-commented type flag for these typed strings and made the necessary changes to support strong type enforcement. For ServerName I did not enable the strong typing yet.

Snarling committed:
Fix test error

d0sboots commented on src/utils/helpers/ObjectHelpers.ts:
```diff
@@ -0,0 +1,4 @@
+/** A wrapper for Object.prototype.hasOwnProperty to avoid calling it from objects directly (lint rule). */
+export function hasOwnProp<T extends object>(obj: T, prop: string): prop is keyof T & string {
+  return Object.prototype.hasOwnProperty.call(obj, prop);
```
`Object.hasOwn` is the preferred replacement. (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwn) It eliminates the need for this helper method. It has wide support, and if it isn't wide enough yet, Babel will transpile it for us.

Snarling replied:
Done

Snarling committed:
Replace hasOwnProp with Object.hasOwn

Also incremented to es2022 to support this

Snarling committed:
Aliases as map, fix test

Changed aliases to use maps while addressing 2 of the 936 outstanding lint failures for the next rule. But aliases should now be completely safe even if there is improperly formatted data in the player save.
Also fixed the test fail.

Snarling committed:
Autolint changes from @typescript-eslint/strict

Enabled the rule, ran linter, disabled rule, reviewed all changes that were automatically made. A few had to be reverted, and did some other manual cleanup afterwards.

Added TruthyFilter array helper (filters an array using filter(Boolean) but with the correct type output).

Redid how disabling routing works in GameRoot - instead of a wrapper function that bypasses typechecking and prevents routing when disabled, early returns are added to each routing function separately.

Snarling committed:
Lint fix

d0sboots commented on src/Augmentation/AugmentationHelpers.tsx:
```diff
@@ -41,16 +40,12 @@ function createAugmentations(): void {
 function initAugmentations(): void {
   resetFactionAugmentations();
-  clearObject(StaticAugmentations);
+  for (const augName of Object.getOwnPropertyNames(StaticAugmentations)) delete StaticAugmentations[augName];
```
Mental note: Good candidate for future JSONMap

Snarling replied:
Yeah I think StaticAugmentations isn't even from the save file, so it can just be a regular map I think.

The getOwnPropertyNames thing here is weird, I was just keeping the functionality the same from the removed helper function. Iterating through keys should be functionally identical in this case though. But yeah we should just redo it as a map soon.

d0sboots commented on src/data/codingcontracttypes.ts:
```diff
@@ -270,14 +269,16 @@ export const codingContractTypesMetadata: ICodingContractTypeMetadata[] = [
           ++k;
         }
         if (++u > d) {
-          break;
```
What was wrong with the breaks here? I liked the old way better.

Snarling replied:
Linter didn't like us using a constant as the condition for a loop, this was my attempt at satisfying that. I'm not a huge fan of it either though.

d0sboots reviewed (approved):
Looks good. I'd submit this before it gets any bigger.
